File: ui-automation-framework/pages/checkout.page.ts
import type { Locator, Page } from '@playwright/test';
import { HeaderComponent } from '../components/header.component';
import { InventoryItemComponent } from '../components/inventory-item.component';
import type { CheckoutCustomer } from '../data/checkout';

/**
 * SauceDemo's checkout is a three-step wizard on consecutive URLs. The steps
 * share navigation and never appear independently, so one page object with
 * clearly grouped locators is easier to maintain than three near-empty classes.
 */
export class CheckoutPage {
  readonly header: HeaderComponent;

  // Step one: customer information
  readonly firstNameInput: Locator;
  readonly lastNameInput: Locator;
  readonly postalCodeInput: Locator;
  readonly continueButton: Locator;
  readonly cancelButton: Locator;
  readonly errorMessage: Locator;

  // Step two: overview
  readonly overviewItems: Locator;
  readonly subtotalLabel: Locator;
  readonly taxLabel: Locator;
  readonly totalLabel: Locator;
  readonly paymentInfo: Locator;
  readonly shippingInfo: Locator;
  readonly finishButton: Locator;

  // Completion
  readonly completeHeader: Locator;
  readonly completeText: Locator;
  readonly backHomeButton: Locator;

  constructor(readonly page: Page) {
    this.header = new HeaderComponent(page);

    this.firstNameInput = page.getByTestId('firstName');
    this.lastNameInput = page.getByTestId('lastName');
    this.postalCodeInput = page.getByTestId('postalCode');
    this.continueButton = page.getByTestId('continue');
    this.cancelButton = page.getByTestId('cancel');
    this.errorMessage = page.getByTestId('error');

    this.overviewItems = page.getByTestId('inventory-item');
    this.subtotalLabel = page.getByTestId('subtotal-label');
    this.taxLabel = page.getByTestId('tax-label');
    this.totalLabel = page.getByTestId('total-label');
    this.paymentInfo = page.getByTestId('payment-info-value');
    this.shippingInfo = page.getByTestId('shipping-info-value');
    this.finishButton = page.getByTestId('finish');

    this.completeHeader = page.getByTestId('complete-header');
    this.completeText = page.getByTestId('complete-text');
    this.backHomeButton = page.getByTestId('back-to-products');
  }

  overviewItem(productName: string): InventoryItemComponent {
    const root = this.overviewItems.filter({
      has: this.page.getByTestId('inventory-item-name').getByText(productName, { exact: true }),
    });
    return new InventoryItemComponent(root);
  }

  /** Fills only provided fields so negative tests can leave inputs empty. */
  async fillCustomerInformation({ firstName, lastName, postalCode }: Partial<CheckoutCustomer>) {
    if (firstName) {
      await this.firstNameInput.fill(firstName);
    }
    if (lastName) {
      await this.lastNameInput.fill(lastName);
    }
    if (postalCode) {
      await this.postalCodeInput.fill(postalCode);
    }
  }

  async submitCustomerInformation(customer: Partial<CheckoutCustomer>): Promise<void> {
    await this.fillCustomerInformation(customer);
    await this.continueButton.click();
  }

  async finishOrder(): Promise<void> {
    await this.finishButton.click();
  }
}

File: ui-tests/src/test/java/saucedemo/pages/CheckoutPage.java
package saucedemo.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import saucedemo.components.Header;
import saucedemo.components.ProductItem;

public class CheckoutPage {
    private final Page page;
    public final Header header;

    public CheckoutPage(Page page) {
        this.page = page;
        this.header = new Header(page);
    }

    // step 1
    public Locator firstName() {
        return page.locator("[data-test='firstName']");
    }

    public Locator lastName() {
        return page.locator("[data-test='lastName']");
    }

    public Locator postalCode() {
        return page.locator("[data-test='postalCode']");
    }

    public Locator error() {
        return page.locator("[data-test='error']");
    }

    public void fillInfo(String first, String last, String postal) {
        if (first != null) {
            firstName().fill(first);
        }
        if (last != null) {
            lastName().fill(last);
        }
        if (postal != null) {
            postalCode().fill(postal);
        }
    }

    public void continueCheckout() {
        page.locator("[data-test='continue']").click();
    }

    public void cancel() {
        page.locator("[data-test='cancel']").click();
    }

    // step 2 - overview
    public Locator overviewItems() {
        return page.locator("[data-test='inventory-item']");
    }

    public ProductItem overviewItem(String name) {
        for (int i = 0; i < overviewItems().count(); i++) {
            Locator one = overviewItems().nth(i);
            if (one.locator("[data-test='inventory-item-name']").innerText().equals(name)) {
                return new ProductItem(one);
            }
        }
        throw new RuntimeException("item not on overview: " + name);
    }

    public Locator subtotal() {
        return page.locator("[data-test='subtotal-label']");
    }

    public Locator tax() {
        return page.locator("[data-test='tax-label']");
    }

    public Locator total() {
        return page.locator("[data-test='total-label']");
    }

    public Locator paymentInfo() {
        return page.locator("[data-test='payment-info-value']");
    }

    public Locator shippingInfo() {
        return page.locator("[data-test='shipping-info-value']");
    }

    public void finish() {
        page.locator("[data-test='finish']").click();
    }

    // complete
    public Locator completeHeader() {
        return page.locator("[data-test='complete-header']");
    }

    public Locator completeText() {
        return page.locator("[data-test='complete-text']");
    }

    public void backHome() {
        page.locator("[data-test='back-to-products']").click();
    }
}
